src/daft-local-execution/src/sinks/grouped_aggregate.rs

```diff
@@ -0,0 +1,436 @@
+    input: &Arc<MicroPartition>,
+    params: &GroupedAggregateParams,
+    high_cardinality_threshold_ratio: f64,
+    partial_agg_threshold: usize,
+    global_strategy_lock: &Arc<Mutex<Option<AggStrategy>>>,
+) -> DaftResult<AggStrategy> {
+    let mut global_strategy = global_strategy_lock.lock().unwrap();
+    // If some other worker has determined a strategy, use that.
+    if let Some(global_strat) = global_strategy.as_ref() {
+        return Ok(global_strat.clone());
+    }
+
+    // Else determine the strategy.
+    let groupby = input.eval_expression_list(params.group_by.as_slice())?;
+
+    let groupkey_hashes = groupby
+        .get_tables()?
+        .iter()
+        .map(|t| t.hash_rows())
+        .collect::<DaftResult<Vec<_>>>()?;
+    let estimated_num_groups = groupkey_hashes
+        .iter()
+        .flatten()
+        .collect::<HashSet<_>>()
+        .len();
+
+    let decided_strategy =
+        if estimated_num_groups as f64 / input.len() as f64 >= high_cardinality_threshold_ratio {
+            AggStrategy::PartitionThenAgg(partial_agg_threshold)
+        } else {
+            AggStrategy::AggThenPartition
+        };
+
+    *global_strategy = Some(decided_strategy.clone());
+    Ok(decided_strategy)
+}
+
+#[derive(Default)]
+struct SinglePartitionAggregateState {
+    partially_aggregated: Vec<MicroPartition>,
+    unaggregated: Vec<MicroPartition>,
+    unaggregated_size: usize,
+}
+
+enum GroupedAggregateState {
+    Accumulating {
+        inner_states: Vec<Option<SinglePartitionAggregateState>>,
+        strategy: Option<AggStrategy>,
+        partial_agg_threshold: usize,
+        high_cardinality_threshold_ratio: f64,
+    },
+    Done,
+}
+
+impl GroupedAggregateState {
+    fn new(
+        num_partitions: usize,
+        partial_agg_threshold: usize,
+        high_cardinality_threshold_ratio: f64,
+    ) -> Self {
+        let inner_states = (0..num_partitions).map(|_| None).collect::<Vec<_>>();
+        Self::Accumulating {
+            inner_states,
+            strategy: None,
+            partial_agg_threshold,
+            high_cardinality_threshold_ratio,
+        }
+    }
+
+    fn push(
+        &mut self,
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+        global_strategy_lock: &Arc<Mutex<Option<AggStrategy>>>,
+    ) -> DaftResult<()> {
+        let Self::Accumulating {
+            ref mut inner_states,
+            strategy,
+            partial_agg_threshold,
+            high_cardinality_threshold_ratio,
+        } = self
+        else {
+            panic!("GroupedAggregateSink should be in Accumulating state");
+        };
+
+        // If we have determined a strategy, execute it.
+        if let Some(strategy) = strategy {
+            strategy.execute_strategy(inner_states, input, params)?;
+            return Ok(());
+        }
+
+        // Otherwise, determine the strategy and execute
+        let decided_strategy = determine_agg_strategy(
```

Comment:
it would be cleaner to have this as a method and it performs the caching of the `decided_strategy`. that way the code could look like 
```
let decided_strategy = self.determine_agg_strategy();
decided_strategy.execute_strategy(inner_states, input, params)?;
```